presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -591,6 +592,8 @@ public PlanOptimizers(
                         .add(new PushRemoteExchangeThroughAssignUniqueId())
                         .add(new InlineProjections(metadata.getFunctionAndTypeManager()))
                         .build()));
+        // Merge join - optimizations to use merge join when data properties meet requirements
+        builder.add(new MergeJoinOptimizer(metadata, sqlParser));
```

Comment:
Essentially MergeJoinOptimizer can only eliminate the local exchange in a JoinNode and convert it to MergeJoinNode

We still rely on AddExchange to first divide the plan to different stages, for example 
JoinNode
-TableScan A ( bucketed by a, sorted by c)
-TableScan B ( bucketed by b, sorted by c)

MergeJoinOptimizer only compares the local sorting properties and see if they're sorted by the join key. Thus it would think it could optimize the plan which is actually wrong because A and B are bucketed by different keys 
We need to rely on AddExchanges to change the plan to 
JoinNode
-Exchange
      -TableScan A ( bucketed by a, sorted by c)
-TableScan B ( bucketed by b, sorted by c)


moving MergeJoinOptimizer will introduce a lot of unnecessary work,

Reply:
Thanks for the explanation!  Makes a lot of sense. Maybe later if we want to expand usage, we can think about how to move it up.